File: lib/triangle.rb
class Triangle
  def initialize(a,b,c)
    check(a,b,c)
  end
  
  def check(a,b,c)
    equilateral = Equilateral.new
    isosceles = Isosceles.new
    scalene = Scalene.new
    non = NotATriangle.new
    
    [non, equilateral, isosceles, scalene].each do |check|
      return check.define() if check.is?(a,b,c)
    end
  end
  
end

class Equilateral
  def is?(a,b,c)
    a==b && a==c
  end
  def define
    :equilateral
  end
end

class Isosceles
  def is?(a,b,c)
    (a==b && a!=c) || (a==c && a!=b) || (b==c && b!=a)
  end
  def define
    :isosceles
  end
end

class Scalene
  def is?(a,b,c)
    a!=b && b!=c
  end
  def define
    :scalene
  end
end

class NotATriangle
  def is?(a,b,c)
    a <= 0 || a + b <= c
  end
  def define
    raise TriangleError, "That's not a triangle, dude."
  end
end

class TriangleError < StandardError
end

File: javascript/src/Triangle.js
class Equilateral {
  constructor(sides) {
    this.sides = sides;
  }
  is() {
    return this.sides[0] == this.sides[1] && this.sides[0] == this.sides[2];
  }
  define() {
    return "equilateral";
  }
}

class Isosceles {
  constructor(sides) {
    this.sides = sides;
  }
  is() {
    return (this.sides[0] == this.sides[1] && this.sides[0] != this.sides[2]) || (this.sides[0] == this.sides[2] && this.sides[0] != this.sides[1]) || (this.sides[1] == this.sides[2] && this.sides[1] != this.sides[0]);
  }
  define() {
    return "isosceles";
  }
}

class Scalene {
  constructor(sides) {
    this.sides = sides;
  }
  is() {
    return this.sides[0] != this.sides[1] && this.sides[1] != this.sides[2];
  }
  define() {
    return "scalene";
  }
}

class TriangleChecker {
  constructor(sides) {
    this.sides = sides;
  }
  run() {
    var equilateral = new Equilateral(sides);
    var isosceles = new Isosceles(sides);
    var scalene = new Scalene(sides);
    var definitions = [equilateral, isosceles, scalene];

    for (var i = 0; i < definitions.length; i++) {
      if (definitions[i].is() == true) {
        return definitions[i].define();
        break;
      }
    };
  };
}